### ISSUE TITLE
Roles not properly parsing

### ISSUE THREAD

**Issue by orchetect** [TheAcharya/MarkersExtractor#66]
In 0.2.5, roles are not being parsed fully in the test project (`BRS_LOCKED6_20210521 - 01 Opening Scene 2023-11-22 07-59-54` - #37).

<img width="319" alt="285124502-659e7bdd-a012-440f-ab11-a899607cc928" src="https://github.com/TheAcharya/MarkersExtractor/assets/11605557/bebb1cca-1744-42b4-8058-e18fb3246888">

<img width="330" alt="285124521-1b7772f5-f06d-4556-b6d8-3f529cbfd150" src="https://github.com/TheAcharya/MarkersExtractor/assets/11605557/dfa1cf5b-4317-41c7-a6b4-881710ca3ef3">

**Comment by orchetect** [comment]
I now have `sync-cli`ps parsing roles correctly, even in very complex projects. It was extremely nuanced when nested clips were involved and took some very careful fine-tuning of the parser to accommodate. MarkersExtractor will now output the roles as they are seen in Final Cut Pro (and therefore as the user expects).

Next one to tackle is multicam clips which are equally nuanced but in different ways.

**Comment by orchetect** [comment]
While we're at it, there are two questions about how we should format data that I've had come up recently while reworking things.

## 1. Title Case for Default Roles

What I'm observing is that sometimes FCP exports roles as lowercase. From what I can tell, it only does that for "built-in" or "default" roles.

For example, FCP will show "Dialogue" but in the FCPXML it will output the role as "dialogue" or "dialogue.dialogue-1".

I have noticed it happen with "Dialogue" and "Music" roles, and I believe "Effects" may be lowercased as well.

(User-defined roles are always verbatim - character case is never changed when output to FCPXML.)

This means that currently, some default roles may output to the manifest file as lowercase since MarkersExtractor is taking them as-is. (The only modification currently being done is reducing the subrole when appropriate. ie: "dialogue.dialogue-1" will be reduced to "dialogue" for the manifest file.)

- **💡 Question:** Do we want to title-case these roles so they appear as they do in FCP? Essentially there would be a process added that detects default role names and makes them title-case (capitalizes first character).

## 2. Matching FCP's Display of Roles: Main vs. Sub-Role

FCP only shows the sub-role in the clip inspector. This is for both audio and video roles. Video roles have an automatic sub-role generated by FCP even though it's not obvious. The audio sub-role is more obvious since it breaks them out and forces you to use them. Small technical detail, but useful to know.

Currently we output both main role and sub-role for video role when it's not reducible. ie: "VFX.Smoke". However, FCP will display this in the inspector view as just its sub-role, ie: "Smoke".

![role-editor](https://github.com/TheAcharya/MarkersExtractor/assets/11605557/ad560715-fd01-43b1-87af-cd3c752efe3e)

![video-role-inspector](https://github.com/TheAcharya/MarkersExtractor/assets/11605557/787781fc-c300-40e7-9b76-452a2b6e3109)

Obviously, outputting main + sub-role provides more data for us to work with.

However, with more recent refactors and bug fixes, I have made audio roles output as FCP displays them in certain cases such as `sync-clip`s where there may be multiple audio sources.

For example, if two audio sources "Dialogue.MixL" and "Dialogue.MixR" are enabled, FCP displays it in the inspector as a comma-separated list of sub-roles, ie: "MixL, MixR". This is what MarkersExtractor will output to the manifest file for `Audio Role`. And this is what I suspect users will expect.

In consideration of making the manifest output consistent, do we also implement this for Video Role to only show sub-role?

If we don't, then do we make Audio Role show main + sub-role for all roles? So instead of "MixL, MixR" it would read as "Dialogue.MixL, Dialogue.MixR". The trouble with this is that it can get very long very quickly. In that complex project you sent me, sometimes there are upwards of 6 audio sources and it is far more concise to use FCP's behavior of only using sub-role.

- **💡 Question:** Do we want to match FCP's display of Roles in the user interface for our output manifest?

**Comment by IAmVigneswaran** [comment]
> Question: Do we want to title-case these roles so they appear as they do in FCP? Essentially there would be a process added that detects default role names and makes them title-case (capitalizes first character).

Cosmetically, it would look nice in the database if we can capitalises first character?

> In consideration of making the manifest output consistent, do we also implement this for Video Role to only show sub-role?

I think it would be highly beneficial for users If we can have `MainRole.Subrole`. It becomes more consistent? Subrole would never be printed or list alone in the manifest output?

Users would have multiple VFX "Stems" for the shots. 

<img width="271" alt="Video-Sub-Roles-Images" src="https://github.com/TheAcharya/MarkersExtractor/assets/118706051/172d4e78-481f-4fc8-9d1e-be90df0757cf">

Hence in the Database's **Video Role & Subrole**  column, users can have `VFX`, `VFX.Mirror Ball`, `VFX.Clean Plate`, `VFX.Main Plate` values in the database. And users can easily create a search filter where  **Video Role & Subrole** contains "VFX".

> If we don't, then do we make Audio Role show main + sub-role for all roles? So instead of "MixL, MixR" it would read as "Dialogue.MixL, Dialogue.MixR". The trouble with this is that it can get very long very quickly. In that complex project you sent me, sometimes there are upwards of 6 audio sources and it is far more concise to use FCP's behavior of only using sub-role.

> Question: Do we want to match FCP's display of Roles in the user interface for our output manifest?

Does it make sense to include all of them?  Example - `Dialogue.MixL,Dialogue.MixR,Dialogue.Boom-1,Dialogue.Boom-1` etc. 

That way, all the values would become a multi-select item in the database. It also gives users insight into their timeline if they have missed anything. Just some thoughts. 

PS: Don't include space when adding ",". 

<details><summary>Using Custom Column Types</summary>
<p>

```shell
csv2notion_neo --token YOUR_TOKEN_HERE --column-types "number,multi_select" test.csv
```
<p align="center"> <img src="https://github.com/TheAcharya/csv2notion-neo/blob/master/assets/example_02.png?raw=true"> </p>

</p>
</details>

**Comment by orchetect** [comment]
> I think it would be highly beneficial for users If we can have `MainRole.Subrole`. It becomes more consistent? Subrole would never be printed or list alone in the manifest output?

Sure, that's doable.

> Does it make sense to include all of them? Example - Dialogue.MixL,Dialogue.MixR,Dialogue.Boom-1,Dialogue.Boom-1 etc.

That would then be consistent, yes. More verbose, but no information is abbreviated or lost.

> Don't include space when adding `,`.

Sure. FCP uses `, ` (with the space) as the delimiter when displaying multiple roles in the UI, so I was just following that convention.

> all the values would become a multi-select item in the database

Clever. Makes sense to format with just a comma separator then.

**Comment by orchetect** [comment]
One possible wrinkle is that FCP allows the `,` comma character to be used when the user names roles. So a consumer of the CSV parsing this field may be more complex than just separating it by comma characters. A possible solution might be to use a null character (ASCII code 0, or `"\n"` in Swift string literal) instad of a comma to separate multiple roles.

For example, the `Audio Role & Subrole` field may contain these three totally valid roles:

```
Dialogue.MixL
Dialogue.MixR
Some Role, With Comma.Some Role, With Comma-1
```

With a comma delimiter, it would be entered in the manifest as follows:

```
Dialogue.MixL,Dialogue.MixR,Some Role, With Comma.Some Role, With Comma-1
```

With a null delimiter, it would be possible to parse it. (Represented as `<NULL>` here for demonstration.)

```
Dialogue.MixL<NULL>Dialogue.MixR<NULL>Some Role, With Comma.Some Role, With Comma-1
```

**Comment by orchetect** [comment]
> drop csv and switch to json

It's not a bad idea, honestly. As the manifest data becomes more descriptive over time, we may outgrow the capabilities of CSV any way.

**Comment by orchetect** [comment]
> I now have `sync-clip`s parsing roles correctly

I have now also resolved Multicam roles parsing issues, and the fix for both will be in 0.2.8.

All roles should be parsing correctly now. That will allow us to move onto other features next-in-line, like role exclusion (#57).

**Comment by orchetect** [comment]
> we can have `MainRole.Subrole`. It becomes more consistent
> include all of them. Example - `Dialogue.MixL,Dialogue.MixR,Dialogue.Boom-1,Dialogue.Boom-1`

Multiple roles are now output to the manifest with full main role + sub-role, using a comma delimiter without a space (`,`). Will be in 0.2.8.
